### PR TITLE
Rewrite for major ⏱  improvements

### PR DESCRIPTION
**Pull request by zwilias**
This is a complete, from-scratch rewrite.

- Uses 24-bit integers to represent the intermediary results, and uses `Bitwise` functions on those to do the heavy lifting. This is _much_ more efficient than encoding every single bit as a union type. 
- Attempt to traverse the whole thing at most twice (could be done in a single pass but it'd be fairly hard to comprehend)
- Since JS and, thus, Elm uses UTF-16 for character encoding, while base64 works best on UTF-8 this includes conversions to and from UTF-16

=>

- [x] memory efficiency
- [x] performance efficiency (~8x faster for encoding, ~6x faster for decoding) (EDIT: now more like ~10 to ~20x for both)
- [x] encodes any Elm string
- [x] decodes anything that can be put in an Elm String. For example, something that would result in `\xD83D` is not accepted.

### CODE REVIEW

**Comment by zwilias** [comment]
Closes #10 and #15

**Comment by zwilias** [comment]
Changed my mind about doing it in a single pass. Twice as fast for decoding, 10x faster for encoding small strings, performance parity for encoding long strings.

🚂  choo choo.

**Comment by zwilias** [comment]
# Decoding

image

# Encoding

image

**Comment by ghost** [comment]
Thank you!

File: src/Base64.elm [hunk 1]
```diff
@@ -1,168 +1,38 @@
-module Base64 exposing (encode, decode)
+module Base64 exposing (decode, encode)
 
-{-| Library for base64 encoding and decoding of Ascii strings.
-For the moment only works with the characters :
+{-| Library for base64 encoding and decoding.
 
-" !\"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~"
+Decodes UTF-16 into UTF-8 for encoding, and decodes UTF-8 into UTF-16 for
+encoding.
```
Comment (BrianHicks) [inline src/Base64.elm:6]: encoding twice here?

Comment (zwilias) [reply src/Base64.elm:6]: 😅  oops.

File: tests/Tests.elm [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            (\( output, input ) ->
+                test ("Can decode '" ++ input ++ "'") <|
+                    \_ ->
+                        decode input
+                            |> Expect.equal (Ok output)
+            )
+        |> describe "decode basics"
+
+
+roundTrip : Test
```
Comment (BrianHicks) [inline tests/Tests.elm:54]: 😻

**Pull request merged**